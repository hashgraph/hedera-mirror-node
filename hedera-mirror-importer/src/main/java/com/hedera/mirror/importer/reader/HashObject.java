package com.hedera.mirror.importer.reader;

/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import java.io.IOException;
import lombok.EqualsAndHashCode;
import lombok.Value;

import com.hedera.mirror.common.domain.DigestAlgorithm;
import com.hedera.mirror.importer.exception.InvalidStreamFileException;

@EqualsAndHashCode(callSuper = true)
@Value
public class HashObject extends AbstractStreamObject {

    int digestType;
    byte[] hash;

    public HashObject(ValidatedDataInputStream vdis, String sectionName, DigestAlgorithm digestAlgorithm) {
        super(vdis);

        try {
            digestType = vdis.readInt(digestAlgorithm.getType(), sectionName, "hash digest type");
            int hashLength = digestAlgorithm.getSize();
            hash = vdis.readLengthAndBytes(hashLength, hashLength, false, sectionName, "hash");
        } catch (IOException e) {
            throw new InvalidStreamFileException(e);
        }
    }

    public HashObject(ValidatedDataInputStream dis, DigestAlgorithm digestAlgorithm) {
        this(dis, null, digestAlgorithm);
    }

    protected HashObject(long classId, int classVersion, int digestType, byte[] hash) {
        super(classId, classVersion);
        this.digestType = digestType;
        this.hash = hash;
    }
}
