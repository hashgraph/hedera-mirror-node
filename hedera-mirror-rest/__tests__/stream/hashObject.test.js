/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import HashObject from '../../stream/hashObject';

describe('HashObject', () => {
  let buffer;
  const classId = BigInt.asIntN(64, BigInt('0xf422da83a251741e'));
  const classVersion = 1;
  const digestType = 0x58ff811b;
  const hash = new Array(48).fill(0xef);

  beforeEach(() => {
    buffer = Buffer.from(
      [].concat(
        [0xf4, 0x22, 0xda, 0x83, 0xa2, 0x51, 0x74, 0x1e], // classId
        [0, 0, 0, 1], // classVersion
        [0x58, 0xff, 0x81, 0x1b], // digest type, sha-384
        [0, 0, 0, 0x30], // length, for sha-384, 48 bytes
        hash // 48-byte hash
      )
    );
  });

  test('getLength', () => {
    const expected = buffer.length;
    const hashObject = new HashObject(buffer);
    expect(hashObject.getLength()).toEqual(expected);
  });

  test('classId', () => {
    const hashObject = new HashObject(buffer);
    expect(hashObject.classId).toEqual(classId);
  });

  test('classVersion', () => {
    const hashObject = new HashObject(buffer);
    expect(hashObject.classVersion).toEqual(classVersion);
  });

  test('digestType', () => {
    const hashObject = new HashObject(buffer);
    expect(hashObject.digestType).toEqual(digestType);
  });

  test('hash', () => {
    const hashObject = new HashObject(buffer);
    expect(hashObject.hash).toEqual(Buffer.from(hash));
  });

  test('truncated buffer', () => {
    expect(() => new HashObject(buffer.slice(0, buffer.length - 4))).toThrowErrorMatchingSnapshot();
  });
});
