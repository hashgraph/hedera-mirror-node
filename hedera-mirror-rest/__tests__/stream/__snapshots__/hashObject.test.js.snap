// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HashObject truncated buffer 1`] = `"Error reading byte array, expect 48-byte data got 44-byte"`;
